### DISCUSSION TITLE
Mounting static files breaks api redirect

### DISCUSSION THREAD

**Discussion by ethan-tonic** [fastapi/fastapi#10458] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
--- main.py
import uvicorn
from fastapi import FastAPI
from fastapi.staticfiles import StaticFiles

from api.api import api_router

api_str = "/api/v1"
app = FastAPI(title="test", openapi_url=f"{api_str}/openapi.json")

app.include_router(api_router, prefix=api_str)

app.mount("/", StaticFiles(directory="static", html=True), name="static")


if __name__ == "__main__":
    uvicorn.run("main:app", host="0.0.0.0", port=8000)

--- api/api.py
from fastapi import APIRouter

from api.endpoints import test_endpoint

api_router = APIRouter()
api_router.include_router(test_endpoint.router, prefix="/test", tags=["test"])

--- api/endpoints/test_endpoint.py
from typing import Any
from fastapi import APIRouter

router = APIRouter()

@router.get("/")
async def root() -> Any:
    return {"message": "Hello World"}
```

### Description

Call the `http://localhost:8000/api/v1/test` endpoint and it will return 404. But if you call `http://localhost:8000/api/v1/test/` it will return correctly. This works both with and without the ending slash if you remove `app.mount`. It seems the presence of `app.mount` breaks the redirect.

### Operating System

Windows

### Operating System Details

Windows 11

### FastAPI Version

0.103.2

### Pydantic Version

2.3.0

### Python Version

3.11.3

**Comment by hasansezertasan** [comment]
I believe Its more of a Starlette issue as far as I can see. I hade the same issue when I mounted a sub application. I checked the source code but couldn't figure out why this is happening.

Edit: @ethan-tonic I figured it out, it's intentional! Take a look at [823](https://github.com/encode/starlette/issues/823)

**Reply by ethan-tonic** [reply]
I think this is a bit different because what they describe there is redirecting from `http://localhost:8000/sub` to `http://localhost:8000/sub/`. Here it doesn't even redirect and just breaks entirely with a 404. Another difference is they're describing the behavior on the mount point itself. Here, the trailing slash breaks even outside the mount point itself.

Based on the link, it seems their reasoning is that without the slash it could accidentally match with another url which isn't valid (e.g. `/sub{...}` when trying to match `/sub`). But I noticed this behavior breaks even when on a route like `/whatever/whatever` even though there's no chance for it to accidentally match something else by accident.

Either way, the behavior of this seems to break a lot when you're trying to serve a web page from your index.

**Reply by hasansezertasan** [reply]
Oh, yeah you're right. My bad, sorry...

**Reply by ethan-tonic** [reply]
I might try mounting two sub applications and seeing if that fixes it. One for the api, one for the static files. That might fix it. I'll report back after I try it. Although, I am not sure if in the past we even needed to do this because almost every reference I saw online to doing this didn't mention having to mount separate sub applications for the api and static files. They only mentioned mounting one for the static files only.

**Reply by ethan-tonic** [reply]
Oh interesting, that fixed it. I think there must be an issue with sub applications matching paths when you set it to the index. I assume what's happening is that when you set static files to serve from the index it automatically classifies every api call as being under that sub application for the static files. Although, I am still not quite sure why it breaks without the trailing slash even when you're not querying the root directly. That seems like it's own issue.
